packages/medusa/src/services/shipping-option.ts

```diff
@@ -0,0 +1,717 @@
+        methodPrice = config.price
+      } else {
+        methodPrice = await this.getPrice_(option, validatedData, config.cart)
+      }
+
+      const toCreate: Partial<ShippingMethod> = {
+        shipping_option_id: option.id,
+        data: validatedData,
+        price: methodPrice,
+      }
+
+      if (config.order) {
+        toCreate.order_id = config.order.id
+      }
+
+      if (config.cart) {
+        toCreate.cart_id = config.cart.id
+      }
+
+      if (config.cart_id) {
+        toCreate.cart_id = config.cart_id
+      }
+
+      if (config.return_id) {
+        toCreate.return_id = config.return_id
+      }
+
+      if (config.order_id) {
+        toCreate.order_id = config.order_id
+      }
+
+      if (config.claim_order_id) {
+        toCreate.claim_order_id = config.claim_order_id
+      }
+
+      const method = await methodRepo.create(toCreate)
+
+      const created = await methodRepo.save(method)
+
+      return methodRepo.findOne({
+        where: { id: created.id },
+        relations: ["shipping_option"],
+      }) as unknown as ShippingMethod
+    })
+  }
+
+  /**
+   * Checks if a given option id is a valid option for a cart. If it is the
+   * option is returned with the correct price. Throws when region_ids do not
+   * match, or when the shipping option requirements are not satisfied.
+   * @param {object} option - the option object to check
+   * @param {Cart} cart - the cart object to check against
+   * @return {ShippingOption} the validated shipping option
+   */
+  validateCartOption(
+    option: ShippingOption,
+    cart: Cart
+  ): ShippingOption | null {
+    if (option.is_return) {
+      return null
+    }
+
+    if (cart.region_id !== option.region_id) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        "The shipping option is not available in the cart's region"
+      )
+    }
+
+    const subtotal = cart.subtotal as number
+
+    const requirementResults: boolean[] = option.requirements.map(
+      (requirement) => {
+        switch (requirement.type) {
+          case "max_subtotal":
+            return requirement.amount > subtotal
+          case "min_subtotal":
+            return requirement.amount <= subtotal
+          default:
+            return true
+        }
+      }
+    )
+
+    if (!requirementResults.every(Boolean)) {
+      throw new MedusaError(
+        MedusaError.Types.NOT_ALLOWED,
+        "The Cart does not satisfy the shipping option's requirements"
+      )
+    }
+
+    return option
+  }
+
+  /**
+   * Creates a new shipping option. Used both for outbound and inbound shipping
+   * options. The difference is registered by the `is_return` field which
+   * defaults to false.
+   * @param {ShippingOption} data - the data to create shipping options
+   * @return {Promise<ShippingOption>} the result of the create operation
+   */
+  async create(data: CreateShippingOptionProps): Promise<ShippingOption> {
```

Comment:
very nit(non-blocking): Naming
```suggestion
  async create(data: CreateShippingOptionProps): Promise<ShippingOption> {
```